File: scripts/drifterfarenheter.py
#!/usr/bin/env python3
"""Drifterfarenheter — detaljanalys av felmönster och driftkvalitet.

Korsrefererar manuella körningar med felkoder, analyserar
energieffektivitet och identifierar ventiler med dold risk.

Krav: Kör trendanalys.py och manuell_analys.py först.

Output:
  - output/drifterfarenheter.json
  - output/drifterfarenheter.csv
"""

import json

import pandas as pd
import numpy as np
from scipy.stats import pearsonr

from common import OUTPUT_DIR, ensure_output_dir


# ---------------------------------------------------------------------------
# Datainläsning
# ---------------------------------------------------------------------------

def load_data():
    """Laddar alla CSV:er som behövs för korsanalys."""
    data = {}
    for name in ["trend_anlaggning", "trend_ventiler", "trend_grenar", "trend_anomalier"]:
        path = OUTPUT_DIR / f"{name}.csv"
        if path.exists():
            data[name.replace("trend_", "")] = pd.read_csv(path)
        else:
            data[name.replace("trend_", "")] = pd.DataFrame()

    for name in ["manuell_analys", "manuell_ventiler"]:
        path = OUTPUT_DIR / f"{name}.csv"
        if path.exists():
            data[name] = pd.read_csv(path)
        else:
            data[name] = pd.DataFrame()

    return data


# ---------------------------------------------------------------------------
# Analys 1: Feltyper som driver manuella ingrepp
# ---------------------------------------------------------------------------

def analyze_manual_vs_errors(data):
    """Korsrefererar manuella körningar med felkoder per ventil."""
    tv = data.get("ventiler", pd.DataFrame())
    mv = data.get("manuell_ventiler", pd.DataFrame())
    if tv.empty or mv.empty:
        return {}

    fel_cols = [c for c in tv.columns if c.startswith("Fel_")]

    # Aggregera feltyper per ventil
    agg = {c: (c, "sum") for c in fel_cols}
    agg["tot_fel"] = ("Totala_fel", "sum")
    fel_per_ventil = tv.groupby("Ventil_ID").agg(**agg).reset_index()

    merged = mv.merge(fel_per_ventil, on="Ventil_ID", how="left")

    # Korrelation: manuella kommandon vs varje feltyp
    korrelationer = {}
    for c in fel_cols:
        if merged[c].sum() > 0 and merged["MAN_totalt"].sum() > 0:
            r, p = pearsonr(merged["MAN_totalt"], merged[c])
            korrelationer[c.replace("Fel_", "")] = {
                "pearson_r": round(r, 3),
                "p_varde": round(p, 4),
                "totalt_antal": int(merged[c].sum()),
            }

    # Rangordna: vilka feltyper driver manuella ingrepp?
    sorted_corr = sorted(korrelationer.items(), key=lambda x: abs(x[1]["pearson_r"]), reverse=True)

    # Riskventiler: hög manuell andel + 100% tillgänglighet (dold risk)
    active = merged[merged["Total_CMD"] > 10]
    risk = active[
        (active["Manuell_andel_%"] > 5) & (active["Medel_tillg"] >= 99.9)
    ].sort_values("Manuell_andel_%", ascending=False)

    risk_ventiler = []
    for _, r in risk.head(10).iterrows():
        dominant_error = ""
        max_err = 0
        for c in fel_cols:
            if r[c] > max_err:
                max_err = r[c]
                dominant_error = c.replace("Fel_", "")

        risk_ventiler.append({
            "ventil": r["Ventil_ID"],
            "gren": int(r["Gren"]),
            "manuell_andel": round(r["Manuell_andel_%"], 1),
            "tillganglighet": round(r["Medel_tillg"], 1),
            "totala_fel": int(r["tot_fel"]),
            "dominerande_fel": dominant_error,
            "dominerande_fel_antal": int(max_err),
        })

    return {
        "korrelationer": dict(sorted_corr),
        "risk_ventiler": risk_ventiler,
        "drivande_feltyp": sorted_corr[0][0] if sorted_corr else None,
        "drivande_korrelation": sorted_corr[0][1]["pearson_r"] if sorted_corr else 0,
    }


# ---------------------------------------------------------------------------
# Analys 2: Energieffektivitet och anomalier
# ---------------------------------------------------------------------------

def analyze_energy_efficiency(data):
    """Analyserar kWh/tömning-mönster och identifierar avvikelser."""
    anl = data.get("anlaggning", pd.DataFrame())
    if anl.empty or "kWh_per_tomning" not in anl.columns:
        return {}

    kwh_t = anl["kWh_per_tomning"]
    energi = anl["Energi_kWh"]
    drifttid = anl["Drifttid_h"]

    best_month = anl.loc[kwh_t.idxmin()]
    worst_month = anl.loc[kwh_t.idxmax()]
    spread = ((kwh_t.max() - kwh_t.min()) / kwh_t.mean()) * 100

    # Korrelation energi vs drifttid
    r_drift, p_drift = pearsonr(energi, drifttid) if len(energi) > 2 else (0, 1)

    # Halvårsanalys
    h1 = kwh_t.iloc[:6].mean()
    h2 = kwh_t.iloc[6:].mean()
    h_change = ((h2 - h1) / h1) * 100 if h1 > 0 else 0

    return {
        "kwh_per_tomning_medel": round(kwh_t.mean(), 2),
        "kwh_per_tomning_min": round(kwh_t.min(), 2),
        "kwh_per_tomning_max": round(kwh_t.max(), 2),
        "basta_manad": best_month["Manad"],
        "samsta_manad": worst_month["Manad"],
        "spridning_pct": round(spread, 1),
        "halvars_forandring_pct": round(h_change, 1),
        "korrelation_energi_drifttid": {
            "r": round(r_drift, 3),
            "p": round(p_drift, 4),
        },
        "total_kwh": round(energi.sum(), 0),
        "total_tomningar": int(anl["Tomningar"].sum()),
    }


# ---------------------------------------------------------------------------
# Analys 3: Manuell andel — trend och mönster
# ---------------------------------------------------------------------------

def analyze_manual_trend(data):
    """Analyserar manuell andel över tid och per gren."""
    man = data.get("manuell_analys", pd.DataFrame())
    mv = data.get("manuell_ventiler", pd.DataFrame())
    if man.empty:
        return {}

    total_man = int(man["MAN_totalt"].sum())
    total_cmd = int(man["Total_CMD"].sum())
    andel = total_man / total_cmd * 100 if total_cmd > 0 else 0

    # Trend: jämför H1 vs H2
    h1 = man.iloc[:6]["Manuell_andel_%"].mean()
    h2 = man.iloc[6:]["Manuell_andel_%"].mean()

    # Toppmanad
    worst = man.loc[man["Manuell_andel_%"].idxmax()]

    result = {
        "total_manuella": total_man,
        "total_kommandon": total_cmd,
        "arsandel_pct": round(andel, 2),
        "h1_andel_pct": round(h1, 2),
        "h2_andel_pct": round(h2, 2),
        "samsta_manad": worst["Manad"],
        "samsta_manad_andel": round(worst["Manuell_andel_%"], 1),
    }

    # Per-gren: vilka grenar har mest manuella körningar?
    if not mv.empty:
        gren_agg = mv.groupby("Gren").agg(
            man_totalt=("MAN_totalt", "sum"),
            cmd_totalt=("Total_CMD", "sum"),
            antal_ventiler=("Ventil_ID", "count"),
        ).reset_index()
        gren_agg["man_andel"] = gren_agg["man_totalt"] / gren_agg["cmd_totalt"] * 100
        gren_agg = gren_agg.sort_values("man_andel", ascending=False)

        top_grenar = []
        for _, r in gren_agg.head(5).iterrows():
            top_grenar.append({
                "gren": int(r["Gren"]),
                "manuell_andel": round(r["man_andel"], 1),
                "manuella": int(r["man_totalt"]),
                "ventiler": int(r["antal_ventiler"]),
            })
        result["grenar_med_hogst_manuell"] = top_grenar

    return result


# ---------------------------------------------------------------------------
# Analys 4: Larm och felmönster
# ---------------------------------------------------------------------------

def analyze_alarm_patterns(data):
    """Analyserar larmmönster och identifierar systematiska problem."""
    anl = data.get("anlaggning", pd.DataFrame())
    tv = data.get("ventiler", pd.DataFrame())
    anomalier = data.get("anomalier", pd.DataFrame())
    if anl.empty:
        return {}

    larm = anl["Larm_totalt"] if "Larm_totalt" in anl.columns else pd.Series()
    if larm.empty:
        return {}

    total = int(larm.sum())
    medel = larm.mean()
    median = larm.median()

    # Jan-spik-analys
    jan_larm = larm.iloc[0]
    resten_medel = larm.iloc[1:].mean()
    jan_faktor = jan_larm / resten_medel if resten_medel > 0 else 0

    # Stabil period (exkl jan)
    stabil = larm.iloc[1:]
    stabil_cv = (stabil.std() / stabil.mean() * 100) if stabil.mean() > 0 else 0

    result = {
        "totala_larm": total,
        "medel_per_manad": round(medel, 0),
        "median_per_manad": round(median, 0),
        "januari_larm": int(jan_larm),
        "januari_faktor_vs_resten": round(jan_faktor, 1),
        "feb_dec_medel": round(resten_medel, 0),
        "feb_dec_variationskoefficient": round(stabil_cv, 1),
    }

    # Anomalier
    if not anomalier.empty:
        larm_anom = anomalier[anomalier["mal"] == "larm_manad"]
        result["antal_anomalier"] = len(larm_anom)

    # Feltypsfördelning
    if not tv.empty:
        fel_cols = [c for c in tv.columns if c.startswith("Fel_")]
        fel_totals = {}
        total_fel = 0
        for c in fel_cols:
            s = int(tv[c].sum())
            fel_totals[c.replace("Fel_", "")] = s
            total_fel += s

        # Andel per feltyp
        fel_med_andel = {}
        for k, v in sorted(fel_totals.items(), key=lambda x: -x[1]):
            fel_med_andel[k] = {
                "antal": v,
                "andel_pct": round(v / total_fel * 100, 1) if total_fel > 0 else 0,
            }
        result["felfordelning"] = fel_med_andel
        result["totala_ventilfel"] = total_fel

        # Ventiler med mekaniska fel (DOES_NOT_OPEN + DOES_NOT_CLOSE)
        mek_cols = [c for c in fel_cols if "OPEN" in c or "CLOSE" in c]
        if mek_cols:
            mek_per_v = tv.groupby("Ventil_ID")[mek_cols].sum().sum(axis=1)
            mek_ventiler = (mek_per_v > 0).sum()
            result["ventiler_med_mekaniska_fel"] = int(mek_ventiler)

    return result


# ---------------------------------------------------------------------------
# Samlad bedömning
# ---------------------------------------------------------------------------

def create_summary(manual_errors, energy, manual_trend, alarms):
    """Sammanfattar huvudfynd."""
    findings = []

    # Manuell vs fel-insikt
    if manual_errors.get("drivande_feltyp"):
        ft = manual_errors["drivande_feltyp"]
        r = manual_errors["drivande_korrelation"]
        findings.append({
            "omrade": "Manuella körningar",
            "fynd": (f"Manuella ingrepp drivs främst av {ft}-fel (r={r:.2f}). "
                     f"Åtgärda dessa feltyper för att minska behovet av manuella körningar."),
            "prioritet": 1 if r > 0.5 else 2,
        })

    if manual_errors.get("risk_ventiler"):
        n = len(manual_errors["risk_ventiler"])
        findings.append({
            "omrade": "Dold risk",
            "fynd": (f"{n} ventiler har hög manuell andel trots 100% tillgänglighet. "
                     "Operatörer kompenserar för automatikproblem — dessa ventiler "
                     "riskerar plötsligt bortfall om operatören missar ingripandet."),
            "prioritet": 1,
        })

    # Energi-insikt
    if energy.get("spridning_pct", 0) > 30:
        findings.append({
            "omrade": "Energieffektivitet",
            "fynd": (f"kWh/tömning varierar {energy['spridning_pct']:.0f}% mellan "
                     f"bästa ({energy['basta_manad']}: {energy['kwh_per_tomning_min']:.1f}) "
                     f"och sämsta ({energy['samsta_manad']}: {energy['kwh_per_tomning_max']:.1f}) månad. "
                     f"H2 var {abs(energy['halvars_forandring_pct']):.0f}% "
                     f"{'bättre' if energy['halvars_forandring_pct'] < 0 else 'sämre'} än H1."),
            "prioritet": 2,
        })

    # Larm-insikt
    if alarms.get("januari_faktor_vs_resten", 0) > 5:
        findings.append({
            "omrade": "Larmanalys",
            "fynd": (f"Januari hade {alarms['januari_faktor_vs_resten']:.0f}x fler larm "
                     f"({alarms['januari_larm']:,}) jämfört med övriga månaders snitt "
                     f"({alarms['feb_dec_medel']:.0f}). Exklusive januari är larmfrekvensen "
                     f"stabil (CV {alarms['feb_dec_variationskoefficient']:.0f}%)."),
            "prioritet": 1,
        })

    if alarms.get("felfordelning"):
        dominant = list(alarms["felfordelning"].items())[0]
        name, info = dominant
        findings.append({
            "omrade": "Felfördelning",
            "fynd": (f"{name} dominerar med {info['andel_pct']:.0f}% av alla ventilfel "
                     f"({info['antal']:,} st). Dessa indikerar att tömningsintervallen "
                     "kan optimeras."),
            "prioritet": 2,
        })

    findings.sort(key=lambda x: x["prioritet"])
    return findings


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def main():
    ensure_output_dir()

    print("Laddar data för drifterfarenhetsanalys...")
    data = load_data()

    print("Analyserar felmönster vs manuella ingrepp...")
    manual_errors = analyze_manual_vs_errors(data)
    if manual_errors.get("korrelationer"):
        for name, info in manual_errors["korrelationer"].items():
            print(f"  {name}: r={info['pearson_r']:.3f}, totalt {info['totalt_antal']}")

    print("Analyserar energieffektivitet...")
    energy = analyze_energy_efficiency(data)
    if energy:
        print(f"  kWh/tömning: {energy['kwh_per_tomning_min']:.1f}–{energy['kwh_per_tomning_max']:.1f} "
              f"(spridning {energy['spridning_pct']:.0f}%)")

    print("Analyserar manuell driftandel...")
    manual_trend = analyze_manual_trend(data)
    if manual_trend:
        print(f"  Årsandel: {manual_trend['arsandel_pct']:.1f}%, "
              f"H1: {manual_trend['h1_andel_pct']:.1f}%, H2: {manual_trend['h2_andel_pct']:.1f}%")

    print("Analyserar larmmönster...")
    alarms = analyze_alarm_patterns(data)
    if alarms:
        print(f"  Totalt: {alarms['totala_larm']:,} larm, "
              f"jan-faktor: {alarms.get('januari_faktor_vs_resten', 0):.0f}x")

    print("Sammanställer fynd...")
    findings = create_summary(manual_errors, energy, manual_trend, alarms)
    print(f"  {len(findings)} huvudfynd identifierade")

    output = {
        "manual_vs_felkoder": manual_errors,
        "energieffektivitet": energy,
        "manuell_trend": manual_trend,
        "larmmonster": alarms,
        "huvudfynd": findings,
    }

    json_path = OUTPUT_DIR / "drifterfarenheter.json"
    with open(json_path, "w", encoding="utf-8") as f:
        json.dump(output, f, ensure_ascii=False, indent=2)
    print(f"\n  Sparad: {json_path}")

    # CSV med riskventiler
    if manual_errors.get("risk_ventiler"):
        risk_df = pd.DataFrame(manual_errors["risk_ventiler"])
        csv_path = OUTPUT_DIR / "drifterfarenheter.csv"
        risk_df.to_csv(csv_path, index=False, encoding="utf-8-sig")
        print(f"  Sparad: {csv_path}")

    print("\n" + "=" * 60)
    print("DRIFTERFARENHETSANALYS KLAR")
    print("=" * 60)


if __name__ == "__main__":
    main()

File: webapp/src/analysis/drifterfarenheter.js
/**
 * Port of scripts/drifterfarenheter.py
 * Cross-correlates manual runs with error codes, energy efficiency, alarm patterns
 */

import { pearsonCorrelation } from '../stats/correlation'

const ERROR_TYPES = [
  'DOES_NOT_CLOSE', 'DOES_NOT_OPEN', 'LEVEL_ERROR',
  'LONG_TIME_SINCE_LAST_COLLECTION', 'ERROR_FEEDBACK_FROM_USER',
]

export function analyzeDrifterfarenheter(trendanalys, ventiler, manuellAnalys, larm) {
  const manualVsErrors = analyzeManualVsErrors(trendanalys, ventiler, manuellAnalys)
  const energy = analyzeEnergyEfficiency(trendanalys)
  const manualTrend = analyzeManualTrend(manuellAnalys)
  const alarms = analyzeAlarmPatterns(trendanalys, ventiler, larm)
  const findings = createSummary(manualVsErrors, energy, manualTrend, alarms)

  return {
    manualVsErrors,
    energy,
    manualTrend,
    alarms,
    findings,
  }
}

function analyzeManualVsErrors(trendanalys, ventiler, manuellAnalys) {
  if (!trendanalys?.valveMonthly?.length || !manuellAnalys?.valveSummary?.length) return {}

  // Aggregate errors per valve
  const errorPerValve = {}
  for (const v of trendanalys.valveMonthly) {
    if (!errorPerValve[v.valveId]) errorPerValve[v.valveId] = { totalErrors: 0, byType: {} }
    errorPerValve[v.valveId].totalErrors += v.totalErrors
    for (const [et, count] of Object.entries(v.errors)) {
      errorPerValve[v.valveId].byType[et] = (errorPerValve[v.valveId].byType[et] || 0) + count
    }
  }

  // Correlations: manual commands vs each error type
  const correlations = {}
  const manData = manuellAnalys.valveSummary
  for (const errType of ERROR_TYPES) {
    const manVals = [], errVals = []
    for (const v of manData) {
      const errs = errorPerValve[v.valveId]
      if (errs) {
        manVals.push(v.manTotal)
        errVals.push(errs.byType[errType] || 0)
      }
    }
    if (manVals.length >= 3 && manVals.some(v => v > 0) && errVals.some(v => v > 0)) {
      const { r, p } = pearsonCorrelation(manVals, errVals)
      correlations[errType] = {
        pearsonR: r,
        pValue: p,
        totalCount: errVals.reduce((s, v) => s + v, 0),
      }
    }
  }

  // Sort by absolute correlation
  const sortedCorr = Object.entries(correlations)
    .sort((a, b) => Math.abs(b[1].pearsonR) - Math.abs(a[1].pearsonR))

  // Risk valves: high manual % + 100% availability
  const riskValves = []
  for (const v of manData) {
    if (v.totalCmd <= 10 || v.manualPct <= 5 || (v.avgAvailability != null && v.avgAvailability < 99.9)) continue
    const errs = errorPerValve[v.valveId] || { totalErrors: 0, byType: {} }
    let dominantError = '', maxErr = 0
    for (const [et, count] of Object.entries(errs.byType)) {
      if (count > maxErr) { maxErr = count; dominantError = et }
    }
    riskValves.push({
      valve: v.valveId,
      branch: v.branch,
      manualPct: v.manualPct,
      availability: v.avgAvailability,
      totalErrors: errs.totalErrors,
      dominantError,
      dominantErrorCount: maxErr,
    })
  }
  riskValves.sort((a, b) => b.manualPct - a.manualPct)

  return {
    correlations: Object.fromEntries(sortedCorr),
    riskValves, // full sorted list
    drivingErrorType: sortedCorr[0]?.[0] || null,
    drivingCorrelation: sortedCorr[0]?.[1]?.pearsonR || 0,
  }
}

function analyzeEnergyEfficiency(trendanalys) {
  const fd = trendanalys?.facilityData
  if (!fd?.length) return {}

  const kwhPerE = fd.map(d => d.kwhPerEmptying)
  const energy = fd.map(d => d.energyKwh)
  const opTime = fd.map(d => d.operationTimeH)

  const bestIdx = kwhPerE.indexOf(Math.min(...kwhPerE))
  const worstIdx = kwhPerE.indexOf(Math.max(...kwhPerE))
  const spread = ((Math.max(...kwhPerE) - Math.min(...kwhPerE)) / avg(kwhPerE)) * 100

  // H1/H2 based on month-of-year, not positional slicing
  const h1Vals = fd.filter(d => d.monthNum <= 6).map(d => d.kwhPerEmptying)
  const h2Vals = fd.filter(d => d.monthNum > 6).map(d => d.kwhPerEmptying)
  const h1 = avg(h1Vals)
  const h2 = avg(h2Vals)
  const hChange = h1 > 0 ? ((h2 - h1) / h1) * 100 : 0

  const corrDrift = energy.length > 2 ? pearsonCorrelation(energy, opTime) : { r: 0, p: 1 }

  return {
    kwhPerEmptyingMean: Math.round(avg(kwhPerE) * 100) / 100,
    kwhPerEmptyingMin: Math.round(Math.min(...kwhPerE) * 100) / 100,
    kwhPerEmptyingMax: Math.round(Math.max(...kwhPerE) * 100) / 100,
    bestMonth: fd[bestIdx]?.month,
    worstMonth: fd[worstIdx]?.month,
    spreadPct: Math.round(spread * 10) / 10,
    halfYearChangePct: Math.round(hChange * 10) / 10,
    corrEnergyDrifttime: { r: corrDrift.r, p: corrDrift.p },
    totalKwh: Math.round(energy.reduce((s, v) => s + v, 0)),
    totalEmptyings: fd.reduce((s, d) => s + d.emptyings, 0),
  }
}

function analyzeManualTrend(manuellAnalys) {
  if (!manuellAnalys?.monthly?.length) return {}

  const m = manuellAnalys.monthly
  const h1 = avg(m.filter(r => r.monthNum <= 6).map(r => r.manualPct))
  const h2 = avg(m.filter(r => r.monthNum > 6).map(r => r.manualPct))
  const worst = m.reduce((w, r) => r.manualPct > w.manualPct ? r : w, m[0])

  return {
    totalManual: manuellAnalys.totalMan,
    totalCommands: manuellAnalys.totalAll,
    yearPct: manuellAnalys.yearPct,
    h1Pct: Math.round(h1 * 100) / 100,
    h2Pct: Math.round(h2 * 100) / 100,
    worstMonth: worst.month,
    worstMonthPct: worst.manualPct,
    topBranches: manuellAnalys.branchSummary.map(b => ({
      branch: b.branch,
      manualPct: b.manualPct,
      manTotal: b.manTotal,
      valveCount: b.valveCount,
    })),
  }
}

function analyzeAlarmPatterns(trendanalys, ventiler, larm) {
  const fd = trendanalys?.facilityData
  if (!fd?.length) return {}

  const alarmVals = fd.map(d => d.alarms)
  const total = alarmVals.reduce((s, v) => s + v, 0)
  const mean = avg(alarmVals)
  const median = [...alarmVals].sort((a, b) => a - b)[Math.floor(alarmVals.length / 2)]

  // January detection based on monthNum, not position
  const janEntries = fd.filter(d => d.monthNum === 1)
  const nonJanEntries = fd.filter(d => d.monthNum !== 1)
  const janAlarms = janEntries.reduce((s, d) => s + d.alarms, 0)
  const nonJanAlarms = nonJanEntries.map(d => d.alarms)
  const restMean = avg(nonJanAlarms)
  const janFactor = restMean > 0 ? janAlarms / (janEntries.length || 1) / restMean : 0

  const restStd = Math.sqrt(avg(nonJanAlarms.map(v => (v - restMean) ** 2)))
  const restCV = restMean > 0 ? restStd / restMean * 100 : 0

  const result = {
    totalAlarms: total,
    meanPerMonth: Math.round(mean),
    medianPerMonth: Math.round(median),
    januaryAlarms: janAlarms,
    januaryFactorVsRest: Math.round(janFactor * 10) / 10,
    febDecMean: Math.round(restMean),
    febDecCV: Math.round(restCV * 10) / 10,
    anomalyCount: trendanalys.anomalies.filter(a => a.target === 'larm_månad').length,
  }

  // Error type distribution
  if (ventiler?.errorTypeTotals) {
    const totalErr = Object.values(ventiler.errorTypeTotals).reduce((s, v) => s + v, 0)
    result.errorDistribution = Object.entries(ventiler.errorTypeTotals)
      .sort((a, b) => b[1] - a[1])
      .map(([type, count]) => ({
        type,
        count,
        pct: totalErr > 0 ? Math.round(count / totalErr * 1000) / 10 : 0,
      }))
    result.totalValveErrors = totalErr
  }

  return result
}

function createSummary(manualVsErrors, energy, manualTrend, alarms) {
  const findings = []

  if (manualVsErrors?.drivingErrorType) {
    const r = manualVsErrors.drivingCorrelation
    findings.push({
      area: 'Manuella körningar',
      finding: `Manuella ingrepp drivs främst av ${manualVsErrors.drivingErrorType}-fel (r=${r.toFixed(2)}). Åtgärda dessa feltyper för att minska behovet av manuella körningar.`,
      priority: Math.abs(r) > 0.5 ? 1 : 2,
    })
  }

  if (manualVsErrors?.riskValves?.length) {
    findings.push({
      area: 'Dold risk',
      finding: `${manualVsErrors.riskValves.length} ventiler har hög manuell andel trots 100% tillgänglighet. Operatörer kompenserar för automatikproblem — dessa ventiler riskerar plötsligt bortfall om operatören missar ingripandet.`,
      priority: 1,
    })
  }

  if (energy?.spreadPct > 30) {
    findings.push({
      area: 'Energieffektivitet',
      finding: `kWh/tömning varierar ${Math.round(energy.spreadPct)}% mellan bästa (${energy.bestMonth}: ${energy.kwhPerEmptyingMin.toFixed(1)}) och sämsta (${energy.worstMonth}: ${energy.kwhPerEmptyingMax.toFixed(1)}) månad. H2 var ${Math.abs(energy.halfYearChangePct).toFixed(0)}% ${energy.halfYearChangePct < 0 ? 'bättre' : 'sämre'} än H1.`,
      priority: 2,
    })
  }

  if (alarms?.januaryFactorVsRest > 5) {
    findings.push({
      area: 'Larmanalys',
      finding: `Januari hade ${Math.round(alarms.januaryFactorVsRest)}x fler larm (${alarms.januaryAlarms.toLocaleString()}) jämfört med övriga månaders snitt (${alarms.febDecMean}). Exklusive januari är larmfrekvensen stabil (CV ${Math.round(alarms.febDecCV)}%).`,
      priority: 1,
    })
  }

  if (alarms?.errorDistribution?.length) {
    const [top] = alarms.errorDistribution
    findings.push({
      area: 'Felfördelning',
      finding: `${top.type} dominerar med ${top.pct}% av alla ventilfel (${top.count.toLocaleString()} st). Dessa indikerar att tömningsintervallen kan optimeras.`,
      priority: 2,
    })
  }

  findings.sort((a, b) => a.priority - b.priority)
  return findings
}

function avg(arr) {
  if (!arr.length) return 0
  return arr.reduce((s, v) => s + v, 0) / arr.length
}
